### PR TITLE
dvc: optimize all target specific commands to not build graph

### PR DESCRIPTION
**Pull request by Suor**
Should help with #2671 for commands like `push/pull/checkout` when particular stage file is specified directly.

### CODE REVIEW

File: dvc/repo/__init__.py [hunk 1]
```diff
@@ -179,28 +179,34 @@ def collect(self, target, with_deps=False, recursive=False, graph=None):
         import networkx as nx
         from dvc.stage import Stage
 
-        G = graph or self.graph
-
         if not target:
-            return list(G)
+            return list(graph) if graph else self.stages
```
Comment (skshetry) [inline dvc/repo/__init__.py]:
```suggestion
            return list(graph or self.stages)
```

File: tests/unit/repo/test_repo.py [hunk 2]
```diff
@@ -66,3 +67,18 @@ def test_locked(mocker):
...
+
+
+def test_collect_optimization(tmp_dir, dvc, mocker):
+    (stage,) = tmp_dir.dvc_gen("foo", "foo text")
+
+    # Forget cached stages and graph and error out on collection
+    dvc._reset()
+    mocker.patch(
+        "dvc.repo.Repo.stages",
+        property(raiser(Exception("Should not collect"))),
```
Comment (pared) [inline tests/unit/repo/test_repo.py]: +1

**Comment by efiop** [comment]
@Suor Please address review comments.

File: dvc/repo/__init__.py (hunk 1 above)
Comment (Suor) [reply dvc/repo/__init__.py]: No need to make a shallow copy off stages.

**Comment by Suor** [comment]
Should be good to go

**Pull request merged**